### ISSUE TITLE
Walrus operators unrecognised in dictionary initialisation of a class attribute

### ISSUE THREAD

**Issue by masasakano** [python/mypy#8379]
It seems Walrus operators (in Python 3.8+) are not recognised by mypy when used in initialising a dictionary of a class attribute. Here is a test code.

```python
class TryWalrus:
    CKEY = { 
        'KY': (p:='a_'),
         'ANOTHER': p+'content',
        }

print(TryWalrus.CKEY['ANOTHER']) # => a_content
```

This outputs the following:

    $ mypy try_walrus.py
    try_walrus.py:4: error: Name 'p' is not defined
    Found 1 error in 1 file (checked 1 source file)

If the variable `CKEY` is out of the class definition, i.e. if directly in Main, it does not raise an error.

I tested it with Python 3.8.1 and mypy version `0.770+dev.8888b1aedf6d4ba543ae12f538f190ce5e65c20d` (yesterday's head).

**Comment by dhood** [comment]
This will also happen with a list e.g. `l = [x2 := 1, x2 + 2, x2 + 3]` (for this case also it will only happen if it is in the class scope).

I have noticed, however, that it only happens on RE-USE of the variable from the assignment expression within the initialisation. Otherwise the walrus operator is ok.

```
class Example:
    # l = [x2 := 1, x2 + 2, x2 + 3] # not OK
    l = [x2 := 1, 2, 3] # OK
```

**Comment by bunny-therapist** [comment]
I just (2023) ran into this exact problem with
```
mypy 1.4.1 (compiled: yes)
Python 3.11.0
```

**Comment by it176131** [comment]
Any proposed fix—besides `# type: ignore[name-defined]`—for this? I ran into it with the following versions:
```text
Python 3.12.2
mypy   1.10.0
```

**Comment by JelleZijlstra** [comment]
Anyone impacted should feel free to work on a PR to fix this. I haven't looked at the code but I imagine it's a relatively easy fix.
